### PR TITLE
DrivAerML: attention dropout=0.05 at champion config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 4L/512d DrivAerML model uses dropout=0.0 everywhere. Adding light attention dropout (p=0.05) forces attention heads to learn redundant, distributed representations instead of overfitting to dominant local mesh patterns. DrivAerML's surface meshes are highly structured (nearby points strongly correlated) — dropout may prevent the model from memorizing local mesh topology.

**Risk is low:** 0.05 dropout is very conservative. spike (#2995) tested 0.1 on TF/AF/DM but that was cross-dataset and used a different architecture/config. This is a focused DrivAerML-only test at the exact 4L/512d champion config with lighter dropout.

## Instructions

Add `--attn-dropout 0.05` flag to train.py. In the Transolver `nn.MultiheadAttention` calls, set `dropout=args.attn_dropout`.

Run one variant:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --attn-dropout 0.05 \
  --wandb_group dm-attn-dropout
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML champion:**
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B: `bht6h42t` (4L/512d, AdamW lr=5e-4, T_max=30, attn dropout=0.0)
- External: Transolver-3 3.71% | AB-UPT 3.82%

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shouko:

## Results

### DrivAerML attn-dropout=0.05 — NEGATIVE RESULT (training instability)

| Metric | Baseline (no dropout) | This run (attn-dropout=0.05) |
|---|---|---|
| val best surface_rel_l2_pct | **3.997%** @ ep467 | **12.53%** @ ep118 |
| Divergence spikes (>35%) | Rare | **7 spikes in 206 epochs** |
| Spike frequency | N/A | Every 12–43 epochs, accelerating |
| Recovery time per spike | N/A | ~20 epochs each |
| Final state (ep206) | — | 62.3% (catastrophic double-spike) |

**W&B run:** `qjceei69` (shouko/dm-attn-dropout-005, group: dm-attn-dropout)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --attn-dropout 0.05 \
  --agent shouko --wandb-name "shouko/dm-attn-dropout-005" --wandb-group dm-attn-dropout
```

**Peak VRAM:** ~2.7GB (single GPU)

### What happened

**Hypothesis falsified.** Attention dropout=0.05 causes catastrophic training instability on DrivAerML with the T_max=30 cosine schedule.

The model showed healthy early descent (64%→12.53% over 118 epochs), reaching a best of 12.53% at ep118. However, severe divergence spikes occurred repeatedly and with increasing frequency:

| Spike | Epoch | Val % | Gap from prev |
|---|---|---|---|
| 1 | 39 | 60.7% | — |
| 2 | 76 | 51.2% | 37 ep |
| 3 | 119 | 59.1% | 43 ep |
| 4 | 146 | 59.5% | 27 ep |
| 5 | 153 | 58.5% | 7 ep (double-spike) |
| 6 | 193 | 36.6% | 40 ep |
| 7 | 205 | 36.6% | 12 ep |
| 7b | 206 | 62.3% | 1 ep (catastrophic) |

Each recovery took ~20 epochs, meaning **~40% of total training time was spent recovering from spikes** rather than making progress. After the ep146-153 double-spike, the model never recovered back to its 12.53% best. The ep205-206 double-spike pushed the model to 62.3% — worse than epoch 1 — with loss at 0.78.

**Root cause analysis:** The dropout introduces stochastic noise at cosine LR peaks (where optimization is already most volatile). With T_max=30, the LR peaks come every 15 epochs and are high enough to amplify the dropout noise into full divergence. The no-EMA config provides no safety net to preserve good checkpoints through spikes. The combination of dropout + T_max=30 + no-EMA is toxic for DrivAerML training stability.

**Key insight:** The DrivAerML champion config operates at a razor's edge of stability with T_max=30 cosine cycling. Even 5% attention dropout is enough to push it over the edge repeatedly. Regularization via dropout is counterproductive when the optimizer is already near its stability boundary.

### Suggested follow-ups
- attn-dropout=0.05 + EMA=0.999 — EMA could preserve best checkpoints through spikes
- attn-dropout=0.05 + T_max=50 or T_max=100 — longer cosine periods have gentler LR peaks
- attn-dropout=0.05 + grad-clip=1.0 — gradient clipping may prevent the cascade that causes spikes
- attn-dropout=0.01 — much lighter dropout that may not trigger instability
- The "dropout as regularization" approach should be tested with EMA before being dismissed entirely

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3113 because ADVISOR: attention dropout=0.05 reaches only 12.53% (ep118) with 7 divergence spikes over 206 epochs — the dropout+T_max=30+no-EMA combination is fundamentally unstable. Your diagnosis is correct: LR peaks amplify dropout noise with no EMA safety net. The interesting follow-up (attn-dropout WITH EMA+gc) will be tested as a separate experiment. Closing this run..

**Comment by tcapelle** [comment]
STUDENT shouko:

## Final test metrics (run completed, 526 epochs)

| Metric | Value |
|---|---|
| Best val surface_rel_l2_pct | **12.53%** @ ep118 |
| Test surface_rel_l2_pct @ best val | **13.47%** |
| Total epochs trained | 526 (360-min timeout) |
| W&B run | `qjceei69` |

The run continued past the ep206 catastrophic spike and trained to ep526, but the best checkpoint remained at ep118. The model never recovered to sub-12.53% after the spike cascade began. Baseline comparison: 3.997% val / test not reported (baseline W&B: `bht6h42t`).
